hub/util/merge.py

```diff
@@ -0,0 +1,321 @@
+from collections import defaultdict
+from typing import Dict, List, Optional, Set, Tuple
+from hub.core.version_control.commit_diff import CommitDiff
+from hub.core.version_control.commit_node import CommitNode
+from hub.util.diff import get_lowest_common_ancestor, sanitize_commit
+from hub.util.exceptions import (
+    MergeConflictError,
+    MergeMismatchError,
+    MergeNotSupportedError,
+)
+from hub.util.keys import get_sample_id_tensor_key, get_tensor_commit_diff_key
+from hub.util.remove_cache import create_read_copy_dataset
+from hub.util.version_control import auto_checkout, auto_commit, commit
+
+
+def merge(
+    dataset,
+    target_id: str,
+    conflict_resolution: Optional[str] = None,
+    delete_removed_tensors=False,
+):
+    version_state = dataset.version_state
+    commit_node_map = version_state["commit_node_map"]
+
+    auto_checkout(dataset)
+    target_commit_id = sanitize_commit(target_id, version_state)
+    target_commit_id = auto_commit_target_commit(dataset, target_commit_id)
+    target_ds = create_read_copy_dataset(dataset, target_commit_id)
+
+    original_node: CommitNode = version_state["commit_node"]
+    target_node: CommitNode = commit_node_map[target_commit_id]
+    lca_id = get_lowest_common_ancestor(original_node, target_node)
+    if lca_id == target_commit_id:
+        print("No merge needed, target id is an ancestor of the current commit")
+        return
+    lca_node: CommitNode = commit_node_map[lca_id]
+
+    original_tensors: Set[str] = set(dataset.tensors.keys())
+    all_original_tensors: Set[str] = set(dataset._all_tensors_filtered())
+    check_id_tensors_exist(original_tensors, all_original_tensors)
+
+    target_tensors: Set[str] = set(target_ds.tensors.keys())
+    all_target_tensors: Set[str] = set(target_ds._all_tensors_filtered())
+    check_id_tensors_exist(target_tensors, all_target_tensors)
+
+    lca_tensors = get_lca_tensors(dataset, lca_id)
+
+    new_tensors = target_tensors - original_tensors
+    common_tensors = target_tensors & original_tensors
+
+    # present in dataset at lca, but deleted in target
+    target_deleted_tensors = lca_tensors - target_tensors
+
+    # present in dataset at lca, but deleted in original
+    original_deleted_tensors = lca_tensors - original_tensors
+
+    target_diff, _ = dataset.diff(target_commit_id, lca_id, as_dict=True)
+    for tensor in original_deleted_tensors:
+        diff = target_diff.get(tensor, None)
+
+        # either target doesn't have the tensor, no point in creating again
+        # or target has the tensor but it wasn't modified
+        if not diff or not (diff.data_added or diff.data_updated):
+            new_tensors.discard(tensor)
```

Comment:
Is there a way to take the housekeeping above into separate logical functions? hard to read the flow.